docs/guide/search/extend_elasticsearch.md

```diff
@@ -0,0 +1,422 @@
+
+        public function onLocationDocumentCreate(LocationIndexCreateEvent $event): void
+        {
+            $document = $event->getDocument();
+            $document->fields[] = new Field(
+                'custom_field',
+                'Custom field value',
+                new StringField()
+            );
+        }
+
+        public static function getSubscribedEvents(): array
+        {
+            return [
+                ContentIndexCreateEvent::class => 'onContentDocumentCreate',
+                LocationIndexCreateEvent::class => 'onLocationDocumentCreate'
+            ];
+        }
+    }
+    ```
+
+    Remember to register the subscriber as a service:
+
+    ``` yaml
+    services:
+        App\EventSubscriber\CustomIndexDataSubscriber:
+            tags:
+                - { name: kernel.event_subscriber }
+    ```
+
+    ## Manipulating the query
+
+    You can customize the search query before it is executed.
+    To do it, subscribe to `Ibexa\Platform\Contracts\ElasticSearchEngine\Query\Event\QueryFilterEvent`.
+
+    The following example shows how to add an additional Search Criterion to all queries.
+
+    Note that depending on your configuration, this will impact all search queries,
+    including those used for search and content tree in the Back Office.
+
+    ``` php hl_lines="35"
+    <?php
+
+    declare(strict_types=1);
+
+    namespace App\EventSubscriber;
+
+    use eZ\Publish\API\Repository\Values\Content\Query\Criterion\LogicalAnd;
+    use eZ\Publish\API\Repository\Values\Content\Query\Criterion\ObjectStateIdentifier;
+    use Ibexa\Platform\Contracts\ElasticSearchEngine\Query\Event\QueryFilterEvent;
+    use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+
+    final class CustomQueryFilterSubscriber implements EventSubscriberInterface
+    {
+        public function onQueryFilter(QueryFilterEvent $event): void
+        {
+            $query = $event->getQuery();
+
+            $additionalCriteria = new ObjectStateIdentifier('locked');
+
+            if ($query->filter !== null) {
+                $query->filter = $additionalCriteria;
+            } else {
+                // Append Criterion to existing filter
+                $query->filter = new LogicalAnd([
+                    $query->filter,
+                    $additionalCriteria
+                ]);
+            }
+        }
+
+        public static function getSubscribedEvents(): array
+        {
+            return [
+                QueryFilterEvent::class => 'onQueryFilter'
+            ];
+        }
+    }
+    ```
+
+    Remember to register the subscriber as a service:
+
+    ``` yaml
+    services:
+        App\EventSubscriber\CustomQueryFilerSubscriber:
+            tags:
+                - { name: kernel.event_subscriber }
+    ```
+
+    ## Custom Search Criterion
+
+    To provide support for a custom Search Criterion, you need to implement `Ibexa\Platform\Contracts\ElasticSearchEngine\Query\CriterionVisitor`:
```

Comment:
```suggestion
    To provide support for a custom Search Criterion, you need to implement the Criterion visitor:
```
WDYS?

Reply:
I'd leave full signature here and below, for easier browsing (otherwise the reader has to look through the code block to find the right line).